### ISSUE TITLE
Mypy is failing to infer generic type instances from default values

### ISSUE THREAD

**Issue by sandrinr** [python/mypy#15707]
**Bug Report**

Mypy is reporting incompatible default arguments when using generic function types in cases where there should be no ambiguity about the type instance.

Note: this could be a duplicate of #15275

**To Reproduce**

```python
from typing import TypeVar

class A:
    pass


T_A = TypeVar("T_A", bound=A)


def f(a: T_A = A()) -> None:
    pass
```

Playground URL: https://mypy-play.net/?mypy=latest&python=3.11&gist=0e6a08418f29c810eba3ca03719119ed

**Expected Behavior**

Mypy does not report an error and assigns `A` to `T_A` in the case when `f` is called without an `a` argument.

**Actual Behavior**

Mypy is reporting:
```
error: Incompatible default for argument "a" (default has type "A", argument has type "T_A")  [assignment]
```

**Your Environment**

- Mypy version used: 1.4.1 (ran on https://mypy-play.net/)
- Mypy command-line flags: none (ran on https://mypy-play.net/)
- Mypy configuration options from `mypy.ini` (and other config files): none (ran on https://mypy-play.net/)
- Python version used: 3.11 (https://mypy-play.net/)

**Comment by A5rocks** [comment]
Err, just like in the other mentioned issue, this has an actual error in that `T_A` is not necessarily `A`. The other issue is only open because I believe there could be some extra inference added (though now that I think of it I'm not convinced about that).

**Comment by A5rocks** [comment]
More specifically:

```py
from typing import TypeVar

class A:
    pass


T_A = TypeVar("T_A", bound=A)


def f(a: T_A = A()) -> T_A:
    return a

class B(A):
  z: int

x: B = f()  # this is currently accepted, as it should be
x.z  # if mypy allowed `A` to be assigned as a default without error, then this would be buggy
```

**Comment by sandrinr** [comment]
Thank you for the explanation.

I did not know that `x: B = f()` is a correct (in the sense of safety) application of a generic type assignment. I probably would have expected MyPy to complain here that `A` being assigned to `T_A` (default value) is not compatible with `B` being assigned to `T_A` (variable declaration) in that case.

**Comment by A5rocks** [comment]
We can't know how `f` will be used when we define it: this means that `A` being assigned to `T_A` has to consider all the scenarios. This specific one doesn't work because of inference. If you want it to work, you might want a `def f(a: A = A()) -> A` instead.

**Comment by sandrinr** [comment]
OK, then let's close this as "works as intended".
